File: Sources/MoonParser/Literals.swift
import MoonAST
import MoonLexer

func parseLiteral(_ ts: TokenStream) throws -> Literal {
    let start = ts.peek()

    if ts.at(.string) {
        let tok = ts.advance()
        let span = ts.makeSpan(start: start, end: tok)
        return .string(tok.value ?? "", span: span)
    }
    if ts.at(.int) {
        let tok = ts.advance()
        let span = ts.makeSpan(start: start, end: tok)
        return .int(Int(tok.value ?? "0") ?? 0, span: span)
    }
    if ts.at(.float) {
        let tok = ts.advance()
        let span = ts.makeSpan(start: start, end: tok)
        return .float(Double(tok.value ?? "0") ?? 0, span: span)
    }
    if ts.at(.kwTrue) {
        let tok = ts.advance()
        let span = ts.makeSpan(start: start, end: tok)
        return .bool(true, span: span)
    }
    if ts.at(.kwFalse) {
        let tok = ts.advance()
        let span = ts.makeSpan(start: start, end: tok)
        return .bool(false, span: span)
    }

    throw ParseError("Expected literal", line: start.line, column: start.column)
}

func literalSpan(_ lit: Literal) -> MoonAST.Span {
    switch lit {
    case .string(_, let span): return span
    case .int(_, let span): return span
    case .float(_, let span): return span
    case .bool(_, let span): return span
    }
}

File: legacy/packages/parser/src/literals.ts
import type { Literal } from "@moon/ast";
import { ParseError, type TokenStream } from "./token-stream";

export function parseLiteral(ts: TokenStream): Literal {
  const start = ts.peek();

  if (ts.at("STRING")) {
    const tok = ts.advance();
    return { kind: "String", value: tok.value!, span: ts.makeSpan(start, tok) };
  }
  if (ts.at("INT")) {
    const tok = ts.advance();
    return { kind: "Int", value: parseInt(tok.value!, 10), span: ts.makeSpan(start, tok) };
  }
  if (ts.at("FLOAT")) {
    const tok = ts.advance();
    return { kind: "Float", value: parseFloat(tok.value!), span: ts.makeSpan(start, tok) };
  }
  if (ts.at("KW_TRUE")) {
    const tok = ts.advance();
    return { kind: "Bool", value: true, span: ts.makeSpan(start, tok) };
  }
  if (ts.at("KW_FALSE")) {
    const tok = ts.advance();
    return { kind: "Bool", value: false, span: ts.makeSpan(start, tok) };
  }

  throw new ParseError("Expected literal", start.line, start.column);
}